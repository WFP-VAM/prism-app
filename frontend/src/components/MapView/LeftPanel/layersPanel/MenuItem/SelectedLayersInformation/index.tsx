import { Chip, createStyles, makeStyles } from '@material-ui/core';
import { LayerType } from 'config/types';
import { useSafeTranslation } from 'i18n';
import React, { memo, useCallback, useEffect, useState } from 'react';

const useStyles = makeStyles(() =>
  createStyles({
    chipRoot: {
      marginLeft: '3%',
    },
  }),
);

interface SelectedLayersInformationProps {
  selectedCategoryLayers: LayerType[];
}
const SelectedLayersInformation = ({
  selectedCategoryLayers,
}: SelectedLayersInformationProps) => {
  const classes = useStyles();
  const { t } = useSafeTranslation();
  const [informationChipLabel, setInformationChipLabel] = useState<string>(
    selectedCategoryLayers.length.toString(),
  );

  useEffect(() => {
    if (!selectedCategoryLayers.length) {
      return;
    }
    setInformationChipLabel(selectedCategoryLayers.length.toString());
  }, [selectedCategoryLayers.length]);

  const handleChipOnMouseEnter = useCallback(() => {
    setInformationChipLabel(
      `${selectedCategoryLayers.length} ${t('Active Layer(s)')}`,
    );
  }, [selectedCategoryLayers.length, t]);

  const handleChipOnMouseLeave = useCallback(() => {
    setInformationChipLabel(selectedCategoryLayers.length.toString());
  }, [selectedCategoryLayers.length]);

  if (!selectedCategoryLayers.length) {
    return null;
  }
  return (
    <Chip
      onMouseEnter={handleChipOnMouseEnter}
      onMouseLeave={handleChipOnMouseLeave}
      classes={{ root: classes.chipRoot }}
      color="secondary"
      label={informationChipLabel}
    />
  );
};

export default memo(SelectedLayersInformation);
