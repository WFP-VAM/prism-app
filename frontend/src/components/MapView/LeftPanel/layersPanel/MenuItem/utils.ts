import { createStyles, makeStyles } from '@material-ui/core';

export const useLayerMenuItemStyles = makeStyles(() =>
  createStyles({
    root: {
      position: 'inherit',
      backgroundColor: '#FFFF',
    },
    rootSummary: {
      backgroundColor: '#FFFF',
    },
    rootDetails: {
      padding: 0,
    },
    expandIcon: {
      color: 'black',
    },
    summaryContent: {
      alignItems: 'center',
    },
    title: {
      color: 'black',
      fontSize: '14px',
      fontWeight: 600,
    },
    chipRoot: {
      marginLeft: '3%',
    },
  }),
);
