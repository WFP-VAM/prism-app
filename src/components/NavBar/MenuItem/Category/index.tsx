import React from 'react';
import {
  Switch,
  Typography,
  withStyles,
  createStyles,
  WithStyles,
} from '@material-ui/core';

import { LayerType } from '../../../../config/types';

function Category({ classes, title, layers }: CategoryProps) {
  return (
    <div className={classes.container}>
      <Typography variant="body2" className={classes.title}>
        {title}
      </Typography>
      <hr />
      {layers.map(({ id, title: layersTitle }) => (
        <div key={id} className={classes.listContainer}>
          <Switch
            size="small"
            color="default"
            inputProps={{ 'aria-label': 'checkbox with default color' }}
          />{' '}
          <Typography variant="body1">{layersTitle}</Typography>
        </div>
      ))}
    </div>
  );
}

const styles = () =>
  createStyles({
    container: {
      marginBottom: 16,
    },

    listContainer: {
      display: 'flex',
      marginBottom: 8,
    },

    title: {
      fontWeight: 'bold',
      textAlign: 'left',
    },
  });

export interface CategoryProps extends LayerType, WithStyles<typeof styles> {}

export default withStyles(styles)(Category);
