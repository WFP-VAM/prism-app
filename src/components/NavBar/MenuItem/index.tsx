import React, { useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import {
  Button,
  Typography,
  Popover,
  withStyles,
  WithStyles,
  createStyles,
  Theme,
  Switch,
} from '@material-ui/core';

import { MenuItemType, LayerType, TableType } from '../../../config/types';
import { addLayer, removeLayer } from '../../../context/mapStateSlice';
import { loadTable } from '../../../context/tableStateSlice';
import { layersSelector } from '../../../context/mapStateSlice/selectors';

function MenuItem({ classes, title, icon, layersCategories }: MenuItemProps) {
  const [anchorEl, setAnchorEl] = useState<HTMLButtonElement | null>(null);

  const selectedLayers = useSelector(layersSelector);
  const dispatch = useDispatch();

  const handleClick = (event: React.MouseEvent<HTMLButtonElement>) => {
    setAnchorEl(event.currentTarget);
  };

  const handleClose = () => {
    setAnchorEl(null);
  };

  const toggleLayerValue = (prevChecked: boolean, layer: LayerType) => {
    if (prevChecked) {
      dispatch(removeLayer(layer));
    } else {
      dispatch(addLayer(layer));
    }
  };

  const showTableClicked = (table: TableType) => {
    dispatch(loadTable(table.id));
  };

  const open = Boolean(anchorEl);
  const id = open ? 'menu-item-popover' : undefined;

  return (
    <>
      <Button
        className={classes.title}
        onClick={handleClick}
        aria-describedby={id}
      >
        <img className={classes.icon} src={icon} alt={title} />
        <Typography variant="body2">{title}</Typography>
      </Button>

      <Popover
        id={id}
        open={open}
        anchorEl={anchorEl}
        onClose={handleClose}
        className={classes.popover}
        anchorOrigin={{
          vertical: 'bottom',
          horizontal: 'center',
        }}
        transformOrigin={{
          vertical: 'top',
          horizontal: 'center',
        }}
        PaperProps={{
          className: classes.paper,
        }}
      >
        {layersCategories.map(({ title: categoryTitle, layers, tables }) => (
          <div key={categoryTitle} className={classes.categoryContainer}>
            <Typography variant="body2" className={classes.categoryTitle}>
              {categoryTitle}
            </Typography>
            <hr />

            {layers.map(layer => {
              const { id: layerId, title: layerTitle } = layer;
              const selected = Boolean(
                selectedLayers.find(({ id: testId }) => testId === layerId),
              );
              return (
                <div key={layerId} className={classes.layersContainer}>
                  <Switch
                    size="small"
                    color="default"
                    checked={selected}
                    onChange={() => toggleLayerValue(selected, layer)}
                    inputProps={{ 'aria-label': layerTitle }}
                  />{' '}
                  <Typography variant="body1">{layerTitle}</Typography>
                </div>
              );
            })}

            {tables.map(table => (
              <Button
                className={classes.button}
                id={table.title}
                key={table.title}
                onClick={() => showTableClicked(table)}
              >
                <Typography variant="body1">{table.title}</Typography>
              </Button>
            ))}
          </div>
        ))}
      </Popover>
    </>
  );
}

const styles = (theme: Theme) =>
  createStyles({
    button: {
      textTransform: 'none',
    },

    title: {
      margin: '0px 14px',
      textTransform: 'uppercase',
      color: 'inherit',

      '&::after': {
        content: '""',
        display: 'inline-block',
        width: 0,
        height: 0,
        marginLeft: 3.5,
        verticalAlign: 3.5,
        borderTop: '3.5px solid',
        borderRight: '3.5px solid transparent',
        borderBottom: 0,
        borderLeft: '3.5px solid transparent',
      },
    },

    icon: {
      width: 18,
      marginRight: 6,
    },

    popover: {
      marginTop: 8,
    },

    paper: {
      padding: '8px 16px',
      backgroundColor: `${theme.palette.primary.main}f9`,
      borderRadius: theme.shape.borderRadius,
    },

    categoryContainer: {
      marginBottom: 16,
    },

    categoryTitle: {
      fontWeight: 'bold',
      textAlign: 'left',
    },

    layersContainer: {
      display: 'flex',
      marginBottom: 8,
    },
  });

export interface MenuItemProps
  extends MenuItemType,
    WithStyles<typeof styles> {}

export default withStyles(styles)(MenuItem);
