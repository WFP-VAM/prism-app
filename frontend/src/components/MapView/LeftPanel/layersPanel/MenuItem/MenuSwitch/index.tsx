import {
  Accordion,
  AccordionDetails,
  AccordionSummary,
  Chip,
  createStyles,
  Grid,
  makeStyles,
  Typography,
} from '@material-ui/core';
import AddIcon from '@material-ui/icons/Add';
import RemoveIcon from '@material-ui/icons/Remove';
import React, {
  ChangeEvent,
  memo,
  useCallback,
  useEffect,
  useMemo,
  useState,
} from 'react';
import { useSelector } from 'react-redux';
import { LayerType } from 'config/types';
import { useSafeTranslation } from 'i18n';
import { Extent } from 'components/MapView/Layers/raster-utils';
import { layersSelector } from 'context/mapStateSlice/selectors';
import { filterActiveLayers } from 'components/MapView/utils';
import SwitchItems from './SwitchItems';

const useStyles = makeStyles(() =>
  createStyles({
    root: {
      position: 'inherit',
      maxWidth: '100%',
    },
    rootSummary: {
      backgroundColor: '#F5F7F8',
    },
    rootDetails: {
      padding: 0,
      backgroundColor: '#FFFFFF',
    },
    expandIcon: {
      color: '#53888F',
    },
    summaryContent: {
      alignItems: 'center',
    },
    chipRoot: {
      marginLeft: '1.5%',
    },
    title: {
      color: '#53888F',
      fontWeight: 500,
    },
  }),
);

interface MenuSwitchProps {
  title: string;
  layers: LayerType[];
  extent?: Extent;
}

const MenuSwitch = memo(({ title, layers, extent }: MenuSwitchProps) => {
  const { t } = useSafeTranslation();
  const selectedLayers = useSelector(layersSelector);
  const classes = useStyles();
  const [isExpanded, setIsExpanded] = useState(false);

  const handleAccordionExpand = useCallback(
    (event: ChangeEvent<{}>, expanded: boolean) => {
      setIsExpanded(expanded);
    },
    [],
  );

  const selectedInternalLayers = useMemo(() => {
    return selectedLayers.filter(layer => {
      return layers.some(internalLayer => {
        return filterActiveLayers(layer, internalLayer);
      });
    });
  }, [layers, selectedLayers]);

  const [informationChipLabel, setInformationChipLabel] = useState<string>(
    selectedInternalLayers.length.toString(),
  );

  useEffect(() => {
    if (!selectedInternalLayers.length) {
      return;
    }
    setInformationChipLabel(selectedInternalLayers.length.toString());
  }, [selectedInternalLayers.length]);

  const handleChipOnMouseEnter = useCallback(() => {
    setInformationChipLabel(
      `${selectedInternalLayers.length} ${t('Active Layer(s)')}`,
    );
  }, [selectedInternalLayers.length, t]);

  const handleChipOnMouseLeave = useCallback(() => {
    setInformationChipLabel(selectedInternalLayers.length.toString());
  }, [selectedInternalLayers.length]);

  const renderedSelectedLayerInformation = useMemo(() => {
    if (!selectedInternalLayers.length) {
      return null;
    }
    return (
      <Chip
        onMouseEnter={handleChipOnMouseEnter}
        onMouseLeave={handleChipOnMouseLeave}
        classes={{ root: classes.chipRoot }}
        color="secondary"
        label={informationChipLabel}
      />
    );
  }, [
    classes.chipRoot,
    handleChipOnMouseEnter,
    handleChipOnMouseLeave,
    informationChipLabel,
    selectedInternalLayers.length,
  ]);

  return (
    <Accordion
      elevation={0}
      classes={{ root: classes.root }}
      onChange={handleAccordionExpand}
    >
      <AccordionSummary
        expandIcon={isExpanded ? <RemoveIcon /> : <AddIcon />}
        classes={{
          root: classes.rootSummary,
          expandIcon: classes.expandIcon,
          content: classes.summaryContent,
        }}
        aria-controls={title}
        id={title}
      >
        <Typography classes={{ root: classes.title }}>{t(title)}</Typography>
        {renderedSelectedLayerInformation}
      </AccordionSummary>
      <AccordionDetails classes={{ root: classes.rootDetails }}>
        <Grid container direction="column">
          <SwitchItems layers={layers} extent={extent} />
        </Grid>
      </AccordionDetails>
    </Accordion>
  );
});

export default MenuSwitch;
