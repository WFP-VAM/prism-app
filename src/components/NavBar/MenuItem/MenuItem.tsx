import React, { useState, useRef } from 'react';
import {
  Button,
  Typography,
  Popper,
  ClickAwayListener,
  Paper,
  withStyles,
  WithStyles,
  createStyles,
  Theme,
} from '@material-ui/core';

import Category from './Category/Category';
import { CategoryType } from '../../../config/types';

function MenuItem({ classes, title, icon, layersList }: MenuItemProps) {
  const [open, setOpen] = useState(false);
  const anchorRef = useRef<any>();

  return (
    <>
      <Button
        ref={anchorRef}
        className={classes.title}
        onClick={() => setOpen(prevOpen => !prevOpen)}
        aria-controls={open ? 'menu-list-grow' : undefined}
        aria-haspopup="true"
      >
        <img className={classes.icon} src={icon} alt={title} />
        <Typography variant="body2">{title}</Typography>
      </Button>

      <Popper
        open={open}
        anchorEl={anchorRef.current}
        className={classes.popper}
        placement="bottom-start"
        transition
        disablePortal
      >
        <ClickAwayListener onClickAway={() => setOpen(false)}>
          <Paper className={classes.paper}>
            {layersList.map(layers => (
              <Category key={layers.title} {...layers} />
            ))}
          </Paper>
        </ClickAwayListener>
      </Popper>
    </>
  );
}

const styles = (theme: Theme) =>
  createStyles({
    title: {
      margin: '0px 14px',
      textTransform: 'uppercase',
      color: 'inherit',

      '&::after': {
        content: '""',
        display: 'inline-block',
        width: 0,
        height: 0,
        marginLeft: 3.5,
        verticalAlign: 3.5,
        borderTop: '3.5px solid',
        borderRight: '3.5px solid transparent',
        borderBottom: 0,
        borderLeft: '3.5px solid transparent',
      },
    },

    icon: {
      width: 18,
      marginRight: 6,
    },

    popper: {
      marginTop: 8,
    },

    paper: {
      padding: '8px 16px',
      backgroundColor: `${theme.palette.primary.main}f9`,
      borderRadius: 4,
    },
  });

export interface MenuItemProps
  extends CategoryType,
    WithStyles<typeof styles> {}

export default withStyles(styles)(MenuItem);
