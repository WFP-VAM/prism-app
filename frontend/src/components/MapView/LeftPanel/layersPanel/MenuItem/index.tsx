import React, { memo, useCallback, useEffect, useMemo, useState } from 'react';
import {
  createStyles,
  Grid,
  Typography,
  Accordion,
  AccordionDetails,
  AccordionSummary,
  makeStyles,
  Chip,
} from '@material-ui/core';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
import { useSelector } from 'react-redux';
import { LayersCategoryType } from 'config/types';
import MenuSwitch from 'components/MapView/LeftPanel/layersPanel/MenuSwitch';
import { useSafeTranslation } from 'i18n';
import { Extent } from 'components/MapView/Layers/raster-utils';
import { layersSelector } from 'context/mapStateSlice/selectors';
import { filterActiveLayers } from 'components/MapView/utils';

const useStyles = makeStyles(() =>
  createStyles({
    root: {
      position: 'inherit',
    },
    rootSummary: {
      backgroundColor: '#D8E9EC',
    },
    rootDetails: {
      padding: 0,
    },
    expandIcon: {
      color: '#53888F',
    },
    summaryContent: {
      alignItems: 'center',
    },
    chipRoot: {
      marginLeft: '3%',
    },
    title: {
      color: '#53888F',
      fontWeight: 600,
    },
  }),
);

interface MenuItemProps {
  title: string;
  icon: string;
  layersCategories: LayersCategoryType[];
  extent?: Extent;
}

const MenuItem = memo(({ title, layersCategories, extent }: MenuItemProps) => {
  const { t } = useSafeTranslation();
  const selectedLayers = useSelector(layersSelector);
  const classes = useStyles();

  const categoryLayers = useMemo(() => {
    return layersCategories
      .map(layerCategory => {
        return layerCategory.layers;
      })
      .flat();
  }, [layersCategories]);

  const selectedCategoryLayers = useMemo(() => {
    return selectedLayers.filter(layer => {
      return categoryLayers.some(categoryLayer => {
        return filterActiveLayers(layer, categoryLayer);
      });
    });
  }, [categoryLayers, selectedLayers]);

  const [informationChipLabel, setInformationChipLabel] = useState<string>(
    selectedCategoryLayers.length.toString(),
  );

  useEffect(() => {
    if (!selectedCategoryLayers.length) {
      return;
    }
    setInformationChipLabel(selectedCategoryLayers.length.toString());
  }, [selectedCategoryLayers.length]);

  const renderedMenuSwitches = useMemo(() => {
    return layersCategories.map((layerCategory: LayersCategoryType) => (
      <MenuSwitch
        key={layerCategory.title}
        title={layerCategory.title}
        layers={layerCategory.layers}
        extent={extent}
      />
    ));
  }, [extent, layersCategories]);

  const handleChipOnMouseEnter = useCallback(() => {
    setInformationChipLabel(
      `${selectedCategoryLayers.length} ${t('Active Layer(s)')}`,
    );
  }, [selectedCategoryLayers.length, t]);

  const handleChipOnMouseLeave = useCallback(() => {
    setInformationChipLabel(selectedCategoryLayers.length.toString());
  }, [selectedCategoryLayers.length]);

  const renderedSelectedLayerInformation = useMemo(() => {
    if (!selectedCategoryLayers.length) {
      return null;
    }
    return (
      <Chip
        onMouseEnter={handleChipOnMouseEnter}
        onMouseLeave={handleChipOnMouseLeave}
        classes={{ root: classes.chipRoot }}
        color="secondary"
        label={informationChipLabel}
      />
    );
  }, [
    classes.chipRoot,
    handleChipOnMouseEnter,
    handleChipOnMouseLeave,
    informationChipLabel,
    selectedCategoryLayers.length,
  ]);

  return (
    <Accordion elevation={0} classes={{ root: classes.root }}>
      <AccordionSummary
        expandIcon={<ExpandMoreIcon />}
        classes={{
          root: classes.rootSummary,
          expandIcon: classes.expandIcon,
          content: classes.summaryContent,
        }}
        aria-controls={title}
        id={title}
      >
        <Typography classes={{ root: classes.title }}>{t(title)}</Typography>
        {renderedSelectedLayerInformation}
      </AccordionSummary>
      <AccordionDetails classes={{ root: classes.rootDetails }}>
        <Grid container direction="column">
          {renderedMenuSwitches}
        </Grid>
      </AccordionDetails>
    </Accordion>
  );
});

export default MenuItem;
