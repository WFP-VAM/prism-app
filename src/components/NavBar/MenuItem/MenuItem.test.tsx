import React from 'react';
import { render } from '@testing-library/react';
import MenuItem from './MenuItem';

jest.mock('./Category/Category', () => 'mock-Category');

const props = {
  title: 'title',
  icon: 'icon.png',
  layersList: [
    {
      title: 'Category 1',
      layers: [
        {
          id: 'ID',
          title: 'layer',
          serverType: 'wms',
          hasDate: false,
          opacity: 0.5,
          legendText: 'legendText',
        },
      ],
    },
  ],
};

test('examples of some things', () => {
  const { container } = render(<MenuItem {...props} />);
  expect(container).toMatchSnapshot();
});
