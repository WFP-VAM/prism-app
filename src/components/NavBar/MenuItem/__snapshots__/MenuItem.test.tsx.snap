// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders as expected 1`] = `
<div>
  <mock-button
    aria-haspopup="true"
    classname="MenuItem-title-1"
  >
    <img
      alt="title"
      class="MenuItem-icon-2"
      src="icon.png"
    />
    <mock-typography
      variant="body2"
    >
      title
    </mock-typography>
  </mock-button>
</div>
`;
